src/mlpack/methods/reinforcement_learning/environment/frozen_lake.hpp

```diff
@@ -0,0 +1,452 @@
+   *
+   * @return Initial state for each episode.
+   */
+  State InitialSample()
+  {
+    stepsPerformed = 0;
+    boardDescription = generateRandomBoard();
+    return State(nRows, nCols);
+  }
+
+  /**
+   * Initial state representation is the current position at (0, 0) 
+   * and a new random bord is created.
+   * 
+   * @param board The environment board of that the user want to initialize to.
+   * @param nRows The height of the environment board.
+   * @param nCols The width of the environment board.
+   * @return A state in which the height and width of the environment board are 
+   *    limited to nRows and nCols, respectively.
+   */
+  State InitialSample(std::vector<std::vector<char>> board, size_t nRows, size_t nCols)
+  {
+    stepsPerformed = 0;
+    this->nRows = nRows;
+    this->nCols = nCols;
+    boardDescription = board;
+    return State(nRows, nCols);
+  }
+
+  /**
+   * This function checks if the agent has reached the terminal state.
+   *
+   * @param state The desired state.
+   * @return true if state is a terminal state, otherwise false.
+   */
+  bool IsTerminal(const State& state) const
+  { 
+    if (maxSteps != 0 && stepsPerformed >= maxSteps)
+    {
+      Log::Info << "Episode terminated due to the maximum number of steps"
+          "being taken.\n";
+      return true;
+    }
+    else if (boardDescription[state.CurRow()][state.CurCol()] == 'H')
+    {
+      Log::Info << "Episode terminated due to agent falling in the hole.\n";
+      return true;
+    }
+    else if (boardDescription[state.CurRow()][state.CurCol()] == 'G')
+    {
+      Log::Info << "Episode terminated, agent reached the goal.\n";
+      return true;
+    }
+    return false;
+  }
+
+  //! Get the number of steps performed.
+  size_t StepsPerformed() const { return stepsPerformed; }
+
+  //! Get the maximum number of steps allowed.
+  size_t MaxSteps() const { return maxSteps; }
+
+  //! Set the maximum number of steps allowed.
+  size_t& MaxSteps() { return maxSteps; }
+
+  //! Get the board description.
+  std::vector<std::vector<char>> Description() const {return boardDescription;}
+
+  //! Set the board description.
+  std::vector<std::vector<char>>& Description() {return boardDescription;}
+
+ private:
+  /**
+   * Utility function that helps generate random environment board.
+   * 
+   * @return board, a 2D array of characters that hold the board description.
+   *    'S' is the Start Tile, the starting position of the agent.
+   *    'G' is the Goal Tile, the agent can walk into a goal tile to win.
+   *    'F' is a Frozen Tile, the agent can walk on it.
+   *    'H' is a Hole Tile, the agent can fall into it and the game ends.
+   */
+  std::vector<std::vector<char>> generateRandomBoard()
+  {
+    bool valid = false;
+    std::vector<std::vector<char>> Board;
+    //! TODO: Tentative, implement max step to control whether
+    //    the we can never generate a possible board.
```

Comment:
```suggestion
    // the we can never generate a possible board.
```

Just remove some extra spaces, also is the TODO something you like to solve as part of the PR?